File: Tree-shaped Deep Forest/Segment Tree/Q1. Range Sum Query - Mutable/Q1_RangeSumQuery-Mutable.py
class NumArray(object):

    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.nums = nums[:]                # store original array
        self.n = len(nums)                 # length of array
        self.bit = [0] * (self.n + 1)      # Binary Indexed Tree (1-indexed)

        # build the BIT
        for i in range(self.n):
            self._init(i, nums[i])         # initialize BIT with nums values

    def _init(self, i, val):
        """helper to update BIT during initialization"""
        i += 1                             # convert to 1-indexed
        while i <= self.n:
            self.bit[i] += val             # add value to current BIT node
            i += (i & -i)                  # move to next responsible node

    def update(self, index, val):
        """
        :type index: int
        :type val: int
        :rtype: None
        """
        diff = val - self.nums[index]      # find difference from old value
        self.nums[index] = val             # update actual array
        self._init(index, diff)            # update BIT with difference

    def _getSum(self, i):
        """helper to get prefix sum up to index i"""
        s = 0
        i += 1                             # convert to 1-indexed
        while i > 0:
            s += self.bit[i]               # add current BIT value
            i -= (i & -i)                  # move to parent node
        return s

    def sumRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: int
        """
        # range sum = prefix(right) - prefix(left - 1)
        return self._getSum(right) - self._getSum(left - 1)

# Your NumArray object will be instantiated and called as such:
# obj = NumArray(nums)
# obj.update(index,val)
# param_2 = obj.sumRange(left,right)

File: Tree-shaped Deep Forest/Segment Tree/Q1. Range Sum Query - Mutable/Q1_RangeSumQuery-Mutable.java
class NumArray {
    private int[] nums;     // stores original array
    private int[] bit;      // Fenwick Tree array
    private int n;          // size of array

    public NumArray(int[] nums) {
        this.nums = nums;          // store original array
        this.n = nums.length;      // get array length
        this.bit = new int[n + 1]; // create BIT (1-indexed)
        
        // build the Fenwick Tree
        for (int i = 0; i < n; i++) {
            init(i, nums[i]);      // initialize BIT with each number
        }
    }

    // function to update BIT at index i by delta
    private void init(int i, int val) {
        i++; // convert to 1-indexed
        while (i <= n) {
            bit[i] += val; // add value to current node
            i += (i & -i); // move to next index in BIT
        }
    }

    public void update(int index, int val) {
        int diff = val - nums[index]; // find the change in value
        nums[index] = val;            // update the actual array
        init(index, diff);            // update the BIT with the difference
    }

    private int getSum(int i) {
        int sum = 0;
        i++; // convert to 1-indexed
        while (i > 0) {
            sum += bit[i];  // add value from current BIT node
            i -= (i & -i);  // move to parent index
        }
        return sum;
    }

    public int sumRange(int left, int right) {
        // sum in range = prefixSum(right) - prefixSum(left - 1)
        return getSum(right) - getSum(left - 1);
    }
}
